src/hypothesis/searchstrategy/types.py

```diff
@@ -37,6 +37,38 @@ def type_sorting_key(t):
     return issubclass(t, collections.abc.Container)
 
 
+def resolve_forward_reference(thing):
+    """Resolve a forward reference, using typing internals."""
+    # First, we try the obvious: return immediately if it's not a ForwardRef
+    try:
+        from typing import _ForwardRef
+    except ImportError:  # pragma: no cover
+        return thing
+    if type(thing) != _ForwardRef:
+        return thing
+    # If someone has resolved this reference before, we use their result
+    if getattr(thing, '__forward_evaluated__', False):
+        return thing.__forward_value__  # pragma: no cover
+    # This technique worked beautifully until the __forward_frame__ attribute
+    # was removed in Python 3.6
+    if hasattr(thing, '__forward_frame__'):  # pragma: no cover
+        try:
+            return thing._eval_type(
+                thing.__forward_frame__.f_globals,
+                thing.__forward_frame__.f_locals)
+        except NameError:
+            pass
+    # If all else fails, try our current globals.  This probably won't work
+    # but it's worth trying.
+    try:
+        return thing._eval_type(globals(), {})
```

Comment:
This seems super unlikely to work. `globals()` here is the globals of the strategies module. It won't have any of the relevant things in scope will it? It's much more likely to return the wrong thing than it is to do the correct thing.

Reply:
How would you feel about a restricted solution that just handles the case of recursive definitions: When we're creating one from a NamedTuple class definition, push the current definition into a DynamicVariable, then use the current class's `__dict__` as the globals for the eval?